core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -271,31 +272,22 @@ class TransactionStateManager(brokerId: Int,
    * When this broker becomes a follower for a transaction log partition, clear out the cache for corresponding transactional ids
    * that belong to that partition.
    */
-  def removeTransactionsForPartition(partition: Int) {
+  def removeTransactionsForTxnTopicPartition(partitionId: Int) {
     validateTransactionTopicPartitionCountIsStable()
 
-    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
-
-    inLock(stateLock) {
-      ownedPartitions.remove(partition)
-      loadingPartitions.remove(partition)
-    }
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partitionId)
 
     def removeTransactions() {
-      var numTxnsRemoved = 0
-
       inLock(stateLock) {
-        for (transactionalId <- transactionMetadataCache.keys) {
-          if (partitionFor(transactionalId) == partition) {
-            // we do not need to worry about whether the transactional id has any ongoing transaction or not since
-            // the new leader will handle it
-            transactionMetadataCache.remove(transactionalId)
-            numTxnsRemoved += 1
-          }
+        transactionMetadataCache.remove(partitionId) match {
+          case Some(txnMetadataCacheEntry) =>
+            info(s"Removed ${txnMetadataCacheEntry.metadataPerTransactionalId.size} cached transaction metadata for $topicPartition on follower transition")
+
+          case None =>
+            throw new IllegalStateException(s"The metadata cache entry for txn partition $partitionId does not exist while trying to remove it.")
```

Comment:
EDIT: similarly for `addLoadedTransactionsToCache`, but slightly different: we only have one background thread inside `scheduler` and we are checking on `loadingPartitions` so that no requests will be handled while loading so no new entries will be appended to the log. So even we have two loading tasks scheduled,  it is safe for the second loaded metadata to replace the first loaded sub-map via `put`, so if `isDefined` we can just log it.